{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Processing WTCCC Summary Statistics Data\n",
    "The following data was downloaded from EBI-EGA accessions EGAD00000000002-EGAD00000000009\n",
    "Access to data was requested first from https://www.wtccc.org.uk/.\n",
    "The data was decrypted with a key generated by the EGA helpdesk and using the EGA download client.\n",
    "\n",
    "The p-values reported in the paper come from the \"basic_snptest\" files. Each numbered file for each disease is the SNP associations for all SNPs found on that chromosome. Each disease has a 2 or 3 letter disease code. These tables will all be aggregated together into a single table reporting the SNP results. The p-value reported will be the smaller of the 2 values in the 'frequentist_add' or 'frequentist_gen' columns of these files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "wd = '/cellar/users/jkhuang/WTCCC_Download/WTCCC_summary_stats/'\n",
    "outdir = '/cellar/users/jkhuang/Data/Projects/Network_GWAS/Data/WTCCC/WTCCC_SNPs/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Get all the files with SNP summary statistics\n",
    "snp_files = [fn for fn in os.listdir(wd) if '_basic_snptest_' in fn]\n",
    "snp_files.sort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Diseases: ['BD', 'CAD', 'CD', 'HT', 'RA', 'T1D', 'T2D']\n"
     ]
    }
   ],
   "source": [
    "# Get all the disease codes from the file names\n",
    "diseases = []\n",
    "for fn in snp_files:\n",
    "    disease_code = fn.split('_')[6]\n",
    "    if disease_code not in diseases:\n",
    "        diseases.append(disease_code)\n",
    "    else:\n",
    "        pass\n",
    "diseases.sort()\n",
    "print 'Diseases:', diseases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Construct concatenated SNP table and save to file\n",
    "for disease in diseases:\n",
    "    disease_snp_files = [fn for fn in snp_files if disease in fn]\n",
    "    disease_chr_snp_tables = []\n",
    "    for fn in disease_snp_files:\n",
    "        disease_chr_snp_table = pd.read_csv(wd+fn, sep='\\t')\n",
    "        chrom = int(fn.split('_')[-1].split('.')[0])\n",
    "        disease_chr_snp_table['Chromosome'] = chrom\n",
    "        disease_chr_snp_table['P-Value'] = disease_chr_snp_table[['frequentist_add', 'frequentist_gen']].min(axis=1)\n",
    "        disease_chr_snp_table_filt = disease_chr_snp_table[['rsid', 'Chromosome', 'pos', 'frequentist_add', 'frequentist_gen', 'P-Value']]\n",
    "        disease_chr_snp_tables.append(disease_chr_snp_table_filt)\n",
    "    disease_full_snp_table = pd.concat(disease_chr_snp_tables).reset_index(drop=True)\n",
    "    disease_full_snp_table.columns = ['rsID', 'Chromosome', 'Pos', 'Trend P-Value', 'Genotypic P-Value', 'Min P-Value']\n",
    "    disease_full_snp_table_filt = disease_full_snp_table[disease_full_snp_table['rsID']!='---'].reset_index(drop=True)\n",
    "    disease_full_snp_table_filt.to_csv(outdir+disease+'_full_WTCCC_snptest_summary_table.csv', index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Disease: T1D\n",
      "Disease: T2D\n",
      "Disease: HT\n",
      "Disease: BD\n",
      "Disease: CD\n",
      "Disease: RA\n",
      "Disease: CAD\n"
     ]
    }
   ],
   "source": [
    "# Generate bed files to be used for LiftOver for WTCCC data\n",
    "from nbgwas import NBGWAS_snp2gene as snp2gene\n",
    "WTCCC_data_dir = outdir\n",
    "for fn in os.listdir(WTCCC_data_dir):\n",
    "    if fn.endswith('_summary_table.csv'):\n",
    "        disease = fn.split('_')[0]\n",
    "        print 'Disease:', disease\n",
    "        WTCCC_snp_table = snp2gene.load_SNP_pvals(WTCCC_data_dir+fn, delimiter='\\t', header=True, cols='0,1,2,5')\n",
    "        WTCCC_snp_table.columns = ['name', 'chrom_num', 'chromEnd', 'pvalue']\n",
    "        WTCCC_snp_table['chromStart'] = WTCCC_snp_table['chromEnd']-1\n",
    "        WTCCC_snp_table['chrom'] = ['chr'+repr(c) for c in WTCCC_snp_table['chrom_num']]\n",
    "        WTCCC_snp_bed = WTCCC_snp_table[['chrom', 'chromStart', 'chromEnd', 'name']]\n",
    "        WTCCC_snp_bed.to_csv(WTCCC_data_dir+fn[:-4]+'.bed', sep='\\t', header=False, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# After liftover, re-merge data\n",
    "for disease in diseases:\n",
    "    hg19_pos = pd.read_csv(WTCCC_data_dir+disease+'_WTCCC_snplocs_galaxy-LiftOver-hg19.bed', sep='\\t', header=-1)\n",
    "    hg19_pos.columns = ['chrom', 'chromStart', 'chromEnd', 'name']\n",
    "    hg19_pos = hg19_pos.set_index('name')\n",
    "    WTCCC_snp_table = pd.read_csv(WTCCC_data_dir+disease+'_full_WTCCC_snptest_summary_table.csv',\n",
    "                                  delimiter='\\t').set_index('rsID')\n",
    "    hg19_WTCCC_snp_table = pd.concat([hg19_pos, WTCCC_snp_table], axis=1).dropna()\n",
    "    hg19_WTCCC_snp_table_filt = hg19_WTCCC_snp_table[['Chromosome', 'chromEnd', 'Pos', 'Trend P-Value', 'Genotypic P-Value', 'Min P-Value']].reset_index(drop=False)\n",
    "    hg19_WTCCC_snp_table_filt.columns = ['Marker', 'Chromosome', 'hg19 Pos', 'NCBI35 Pos', 'Trend P-Value', 'Genotypic P-Value', 'Min P-Value']\n",
    "    hg19_WTCCC_snp_table_filt['hg19 Pos'] = hg19_WTCCC_snp_table_filt['hg19 Pos'].astype(int)\n",
    "    hg19_WTCCC_snp_table_sorted = hg19_WTCCC_snp_table_filt.sort_values(by=['Min P-Value', 'Chromosome', 'hg19 Pos'])\n",
    "    hg19_WTCCC_snp_table_sorted.to_csv(WTCCC_data_dir+disease+'_WTCCC_snptest_galaxy-LiftOver-hg19_summary_table.csv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Marker</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>hg19 Pos</th>\n",
       "      <th>NCBI35 Pos</th>\n",
       "      <th>Trend P-Value</th>\n",
       "      <th>Genotypic P-Value</th>\n",
       "      <th>Min P-Value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>179987</th>\n",
       "      <td>rs17480050</td>\n",
       "      <td>8</td>\n",
       "      <td>19381917</td>\n",
       "      <td>19426197</td>\n",
       "      <td>1.754650e-214</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.754650e-214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>270213</th>\n",
       "      <td>rs3777582</td>\n",
       "      <td>6</td>\n",
       "      <td>45914028</td>\n",
       "      <td>46022006</td>\n",
       "      <td>1.844440e-20</td>\n",
       "      <td>5.026180e-121</td>\n",
       "      <td>5.026180e-121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46341</th>\n",
       "      <td>rs11042656</td>\n",
       "      <td>11</td>\n",
       "      <td>10214455</td>\n",
       "      <td>10171031</td>\n",
       "      <td>6.606760e-18</td>\n",
       "      <td>6.172960e-111</td>\n",
       "      <td>6.172960e-111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66823</th>\n",
       "      <td>rs11705626</td>\n",
       "      <td>22</td>\n",
       "      <td>25435959</td>\n",
       "      <td>23760513</td>\n",
       "      <td>7.880970e-76</td>\n",
       "      <td>1.399170e-93</td>\n",
       "      <td>1.399170e-93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121222</th>\n",
       "      <td>rs1477523</td>\n",
       "      <td>7</td>\n",
       "      <td>136805138</td>\n",
       "      <td>136262393</td>\n",
       "      <td>3.707340e-06</td>\n",
       "      <td>5.104710e-75</td>\n",
       "      <td>5.104710e-75</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Marker  Chromosome   hg19 Pos  NCBI35 Pos  Trend P-Value  \\\n",
       "179987  rs17480050           8   19381917    19426197  1.754650e-214   \n",
       "270213   rs3777582           6   45914028    46022006   1.844440e-20   \n",
       "46341   rs11042656          11   10214455    10171031   6.606760e-18   \n",
       "66823   rs11705626          22   25435959    23760513   7.880970e-76   \n",
       "121222   rs1477523           7  136805138   136262393   3.707340e-06   \n",
       "\n",
       "        Genotypic P-Value    Min P-Value  \n",
       "179987       1.000000e+00  1.754650e-214  \n",
       "270213      5.026180e-121  5.026180e-121  \n",
       "46341       6.172960e-111  6.172960e-111  \n",
       "66823        1.399170e-93   1.399170e-93  \n",
       "121222       5.104710e-75   5.104710e-75  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hg19_WTCCC_snp_table_sorted.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Constructing NCBI Build 35 gene locations map\n",
    "Constructing this map requires the following files:\n",
    "\n",
    "NCBI Build 35 Known Gene Locations: http://hgdownload.soe.ucsc.edu/goldenPath/hg17/database/knownGene.txt.gz  \n",
    "Known Gene Alias Map: http://hgdownload.soe.ucsc.edu/goldenPath/hg17/database/kgXref.txt.gz  \n",
    "Current HUGO Gene Symbols Table: [ftp://ftp.ebi.ac.uk/pub/databases/genenames/new/tsv/locus_types/gene_with_protein_product.txt](ftp://ftp.ebi.ac.uk/pub/databases/genenames/new/tsv/locus_types/gene_with_protein_product.txt)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "wd2 = '/cellar/users/jkhuang/Data/Projects/Network_GWAS/Data/WTCCC/NCBI_Build35/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Load known gene locations\n",
    "knownGenes = pd.read_csv(wd2+'knownGene.txt', sep='\\t', header=-1)\n",
    "knownGenes.columns = ['Name', 'Chr', 'Strand', 'txStart', 'txEnd', 'cdsStart', 'cdsEnd', 'exonCount', 'exonStart', 'exonEnd', 'UniprotID', 'GencodeID']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 463,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Keep only known gene locations on autosomes\n",
    "autosomes = ['chr'+str(i) for i in range(1,23)]\n",
    "autosome_knownGenes = knownGenes[knownGenes['Chr'].isin(autosomes)]\n",
    "autosome_knownGenes = autosome_knownGenes[['Name', 'Chr', 'Strand', 'txStart', 'txEnd']]\n",
    "autosome_knownGenes = autosome_knownGenes.reset_index(drop=True)\n",
    "# Remove duplicates after only keeping chromosome, strand, and location information\n",
    "autosome_knownGenes_filt = autosome_knownGenes.set_index('Name').drop_duplicates().reset_index(drop=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 465,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "knownGene_names = list(set(autosome_knownGenes_filt['Name']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Load known gene Aliases\n",
    "aliases = pd.read_csv(wd2+'kgXref.txt', sep='\\t', header=-1)\n",
    "aliases.columns=['Gene', 'mRNAID', 'UniprotAccession', 'UniprotID', 'Gene Symbol', 'RefSeq ID', 'Entrez', 'Description']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Load HGNC gene symbol table\n",
    "HGNC_table = pd.read_csv(wd2+'HGNC_Gene_Table_2018-03-20.txt', sep='\\t')\n",
    "\n",
    "# Set all valid HGNC symbols\n",
    "HGNC_symbols = list(HGNC_table['symbol'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 497,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Map all known HGNC aliases\n",
    "HGNC_alias_map = {}\n",
    "for i in HGNC_table.index:\n",
    "    symbol = HGNC_table.ix[i]['symbol']\n",
    "    HGNC_alias_map[symbol]=[symbol]\n",
    "    \n",
    "    # Map gene symbol aliases\n",
    "    if type(HGNC_table.ix[i]['alias_symbol'])==str:\n",
    "        alias_symbols = [name.upper() for name in HGNC_table.ix[i]['alias_symbol'].split('|')]\n",
    "        for alias_symbol in alias_symbols:\n",
    "            if alias_symbol not in HGNC_alias_map:\n",
    "                HGNC_alias_map[alias_symbol]=[symbol]\n",
    "            else:\n",
    "                if symbol not in HGNC_alias_map[alias_symbol]:\n",
    "                    HGNC_alias_map[alias_symbol].append(symbol)\n",
    "    else:\n",
    "        pass\n",
    "    \n",
    "    # Map previous gene symbol names\n",
    "    if type(HGNC_table.ix[i]['prev_symbol'])==str:\n",
    "        prev_symbol = HGNC_table.ix[i]['prev_symbol']\n",
    "        if prev_symbol not in HGNC_alias_map:\n",
    "            HGNC_alias_map[prev_symbol] = [symbol]\n",
    "        else:\n",
    "            HGNC_alias_map[alias_symbol].append(symbol)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 534,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Clean up HGNC aliases to point at only itself\n",
    "for gene in HGNC_alias_map:\n",
    "    if (gene in HGNC_alias_map[gene]) & (len(HGNC_alias_map[gene]) > 1):\n",
    "        HGNC_alias_map[gene] = [gene]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 568,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NM_001010890\n",
      "NM_001004060\n",
      "NM_001004067\n",
      "NM_001786\n",
      "NM_173490\n",
      "NM_001003702\n",
      "NM_032308\n",
      "NM_006876\n",
      "NM_005324\n",
      "NM_003910\n",
      "NM_181538\n",
      "NM_015576\n",
      "NM_033546\n",
      "NM_020640\n",
      "NM_178449\n",
      "NM_178445\n",
      "NM_003616\n",
      "NM_002107\n",
      "NM_080820\n",
      "NM_001005287\n",
      "BT007314\n",
      "NM_018006\n",
      "NM_152250\n",
      "NM_152251\n",
      "NM_030663\n",
      "NM_002567\n",
      "NM_033377\n",
      "NM_001001317\n",
      "NM_014226\n",
      "NM_152499\n",
      "NM_016283\n",
      "NM_181809\n",
      "NM_001004697\n",
      "NM_145047\n",
      "NM_018556\n",
      "NM_016532\n",
      "NM_000204\n",
      "CR608316\n",
      "NM_182831\n",
      "NM_000398\n",
      "NM_144565\n",
      "NM_001091\n",
      "NM_017979\n",
      "NM_014727\n",
      "NM_175054\n",
      "NM_006250\n",
      "NM_015089\n",
      "NM_024311\n",
      "NM_177926\n",
      "NM_003299\n",
      "NM_001004136\n",
      "NM_022452\n",
      "NM_021182\n",
      "AK021680\n",
      "NM_003482\n",
      "BC001926\n",
      "NM_207355\n",
      "NM_006067\n",
      "NM_004038\n",
      "NM_018671\n",
      "NM_024888\n",
      "NM_001006109\n",
      "AB023508\n",
      "BC061588\n",
      "NM_020755\n",
      "AB037703\n",
      "NM_001009610\n",
      "NM_006460\n",
      "AK223365\n",
      "NM_020536\n",
      "NM_024021\n",
      "NM_003537\n",
      "NM_003531\n",
      "NM_001001786\n",
      "NM_032495\n",
      "BC018650\n",
      "NM_198428\n",
      "NM_003536\n",
      "NM_003534\n",
      "NM_003535\n",
      "NM_003532\n",
      "NM_003533\n",
      "NM_003530\n",
      "NM_003538\n",
      "NM_003539\n",
      "NM_002665\n",
      "NM_000344\n",
      "NM_057159\n",
      "NM_176870\n",
      "NM_004942\n",
      "AK098544\n",
      "NM_004084\n",
      "NM_031950\n",
      "NM_000558\n",
      "NM_033542\n",
      "NM_032038\n",
      "NM_012292\n",
      "NM_001008747\n",
      "NM_144621\n",
      "NM_181773\n",
      "NM_003548\n",
      "NM_003546\n",
      "NM_003545\n",
      "NM_003544\n",
      "NM_003543\n",
      "NM_003542\n",
      "NM_003541\n",
      "NM_003540\n",
      "NM_005254\n",
      "NM_020683\n",
      "NM_145175\n",
      "NM_005042\n",
      "NM_001008662\n",
      "NM_007018\n",
      "NM_016100\n",
      "NM_018835\n",
      "NM_016522\n",
      "NM_002307\n",
      "NM_017632\n",
      "NM_018116\n",
      "NM_018117\n",
      "NM_002485\n",
      "NM_023068\n",
      "NM_002389\n",
      "NM_014267\n",
      "BC069285\n",
      "NM_177458\n",
      "NM_003518\n",
      "NM_015444\n",
      "NM_174911\n",
      "NM_004882\n",
      "NM_021968\n",
      "NM_021967\n",
      "BC040528\n",
      "NM_003495\n",
      "NM_014618\n",
      "NM_002036\n",
      "NM_020990\n",
      "NM_170695\n",
      "BC041803\n",
      "NM_016570\n",
      "U11863\n",
      "NM_018661\n",
      "NM_002727\n",
      "NM_024013\n",
      "NM_006295\n",
      "NM_025009\n",
      "AL136719\n",
      "NM_080389\n",
      "NM_080386\n",
      "NM_003525\n",
      "NM_003526\n",
      "NM_003523\n",
      "NM_003522\n",
      "NM_003529\n",
      "NM_032906\n",
      "NM_000322\n",
      "NM_001296\n",
      "NM_015909\n",
      "NM_017411\n",
      "NM_006025\n",
      "NM_020179\n",
      "NM_007236\n",
      "NM_014485\n",
      "NM_014489\n",
      "NM_001001417\n",
      "NM_214710\n",
      "NM_198947\n",
      "NM_030623\n",
      "NM_006188\n",
      "NM_017516\n",
      "BC038989\n",
      "NM_021018\n",
      "NM_012385\n",
      "BC022832\n",
      "NM_016086\n",
      "NM_000805\n",
      "NM_015369\n",
      "NM_005054\n",
      "NM_182919\n",
      "NM_006713\n",
      "BC071590\n",
      "NM_014679\n",
      "NM_018396\n",
      "NM_178439\n",
      "NM_182493\n",
      "NM_006900\n",
      "NM_025146\n",
      "BC053536\n",
      "NM_030772\n",
      "NM_005879\n",
      "NM_000517\n",
      "NM_006588\n",
      "NM_001007553\n",
      "AY598439\n",
      "NM_130384\n",
      "NM_017897\n",
      "NM_005981\n",
      "NM_006982\n",
      "NM_006471\n",
      "NM_017701\n"
     ]
    }
   ],
   "source": [
    "multiple_chromosomes = []\n",
    "failed_mappings = []\n",
    "geneloc_mapping = []\n",
    "multiple_mappings = []\n",
    "for gene in knownGene_names:\n",
    "    genelocs = autosome_knownGenes_filt[autosome_knownGenes_filt['Name']==gene]\n",
    "    # Get gene locations\n",
    "    if len(set(genelocs['Chr']))==1:\n",
    "        chrom = int(list(set(genelocs['Chr']))[0][3:])\n",
    "        TxStarts, TxEnds = [], []\n",
    "        for i in genelocs.index:\n",
    "            if genelocs.ix[i]['Strand']=='+':\n",
    "                TxStarts.append(genelocs.ix[i]['txStart'])\n",
    "                TxEnds.append(genelocs.ix[i]['txEnd'])\n",
    "            else:\n",
    "                TxStarts.append(genelocs.ix[i]['txEnd'])\n",
    "                TxEnds.append(genelocs.ix[i]['txStart'])\n",
    "        TxStart, TxEnd = min(TxStarts), max(TxEnds)\n",
    "    else:\n",
    "        multiple_chromosomes.append(gene)\n",
    "        continue\n",
    "    \n",
    "    # Get all aliases\n",
    "    gene_aliases = aliases[aliases['Gene']==gene]\n",
    "    gene_symbols = list(gene_aliases['Gene Symbol'].dropna())\n",
    "    # Map all alias symbols to official symbol\n",
    "    official_symbols = []\n",
    "    for symbol in gene_symbols:\n",
    "        if symbol in HGNC_alias_map:\n",
    "            official_symbols = official_symbols + HGNC_alias_map[symbol]\n",
    "        else:\n",
    "            official_symbols.append(symbol)\n",
    "    HGNC_genes_1 = HGNC_table[HGNC_table['symbol'].isin(list(set(official_symbols)))]\n",
    "    # Map any refseq ids to HGNC_table\n",
    "    refseq_ids = list(gene_aliases['RefSeq ID'].dropna())\n",
    "    HGNC_genes_2 = HGNC_table[HGNC_table['refseq_accession'].isin(list(set(refseq_ids)))]\n",
    "    # Map any uniprot ids to HGNC_table\n",
    "    uniprot_ids = list(gene_aliases['UniprotAccession'].dropna()) + list(gene_aliases['UniprotID'].dropna())\n",
    "    HGNC_genes_3 = HGNC_table[HGNC_table['uniprot_ids'].isin(list(set(uniprot_ids)))]\n",
    "    \n",
    "    # Concatenate all results together\n",
    "    HGNC_genes = pd.concat([HGNC_genes_1, HGNC_genes_2, HGNC_genes_3]).drop_duplicates()\n",
    "    if HGNC_genes.shape[0]==0:\n",
    "        failed_mappings.append(gene)\n",
    "        continue\n",
    "    elif HGNC_genes.shape[0]==1:\n",
    "        gene_symbol = list(HGNC_genes['symbol'])[0]\n",
    "        geneloc_mapping.append([gene, gene_symbol, chrom, TxStart, TxEnd])\n",
    "    else:\n",
    "        print gene\n",
    "        if gene.startswith('NM_'):\n",
    "            mapped_symbol_list = list(HGNC_genes[HGNC_genes['refseq_accession']==gene]['symbol'])\n",
    "            if len(mapped_symbol_list)==1:\n",
    "                gene_symbol = mapped_symbol_list[0]\n",
    "                geneloc_mapping.append([gene, gene_symbol, chrom, TxStart, TxEnd])\n",
    "            else:\n",
    "                multiple_mappings.append([gene, HGNC_genes[['symbol', 'alias_symbol', 'prev_symbol', 'uniprot_ids', 'refseq_accession']]])\n",
    "        else:\n",
    "            mapped_symbol_list = list(HGNC_genes[HGNC_genes['uniprot_ids']==gene]['symbol'])\n",
    "            if len(mapped_symbol_list)==1:\n",
    "                gene_symbol = mapped_symbol_list[0]\n",
    "                geneloc_mapping.append([gene, gene_symbol, chrom, TxStart, TxEnd])\n",
    "            else:\n",
    "                multiple_mappings.append([gene, HGNC_genes[['symbol', 'alias_symbol', 'prev_symbol', 'uniprot_ids', 'refseq_accession']]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 584,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "mapped_symbols = list(set(NCBI_Build35_genelocs['Symbol']))\n",
    "NCBI_Build35_genelocs_collapsed = []\n",
    "for gene in mapped_symbols:\n",
    "    if len(set(NCBI_Build35_genelocs[NCBI_Build35_genelocs['Symbol']==gene]['Chromosome'])) == 1:\n",
    "        genelocs = NCBI_Build35_genelocs[NCBI_Build35_genelocs['Symbol']==gene]\n",
    "        chrom = list(set(genelocs['Chromosome']))[0]\n",
    "        txStart = genelocs['TxStart'].min()\n",
    "        txEnd = genelocs['TxEnd'].max()\n",
    "        NCBI_Build35_genelocs_collapsed.append([gene, chrom, txStart, txEnd])\n",
    "NCBI_Build35_genelocs_table = pd.DataFrame(NCBI_Build35_genelocs_collapsed, columns=['Gene', 'Chromosome', 'TxStart', 'TxEnd'])        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 588,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "NCBI_Build35_genelocs_table = NCBI_Build35_genelocs_table.sort_values(by=['Chromosome', 'TxStart'], ascending=[True, True])\n",
    "NCBI_Build35_genelocs_table.to_csv('/cellar/users/jkhuang/Data/Projects/Network_GWAS/Data/WTCCC/NCBI_Build35/NCBI_Build35_HGNC_genelocs.csv', \n",
    "                                   sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
